SolidJS widget to format date[time]s

acarapetis opened the pull request:
Here's my attempt at #42: a SolidJS component FormatDate that can render as an absolute date, absolute datetime or a relative time ago. I've replaced all usage of TimeAgo with it, and also updated MatchPreview to use it.

I've got one annoyign bug hanging around - when using a non-relative format, the rendered HTML is not updating currently when hydrating on the client - the UTC times just hang around forever. The `console.log` debugging call I've got in there is outputting what I expect on the client, so the code is definitely running; it's just not rendering correctly for some reason.

This is my first time touching both Astro and SolidJS so I figured I'd ask for help rather than spend any more time banging my head against it - there's clearly something I don't understand either in how the SSR hydration works, or just how SolidJS works.  Also open to any other feedback :)

acarapetis committed:
WIP: solidjs widget to format date[time]s

robertvanhoesel committed:
Replace isServer with onMount

robertvanhoesel commented on src/components/widgets/FormatDate.tsx:
```diff
@@ -0,0 +1,101 @@
+  const [output, setOutput] = createSignal<string>(formatDate(d, format, true))
+  const isRelative = format === "relative" || format === "shortrelative"
+
+  onMount(() => {
```
I replaced the `isServer` boolean with `onMount`, a hook that will only be called once, and only client side. I'm not sure if `isServer` is fully supported in Astro, but either way you want to only call these functions in a browser context.

I left your code structure as is but can be simplified after testing a bit more.

acarapetis replied:
Looks like that's fixed the bug, thanks!

robertvanhoesel commented on src/components/widgets/FormatDate.tsx:
```diff
@@ -0,0 +1,101 @@
+}
+export default function FormatDate({ date, format }: FormatDateProps) {
+  const d = parseUTC(date)
+  const [output, setOutput] = createSignal<string>(formatDate(d, format, true))
```
We can initialize here at once

robertvanhoesel commented on src/components/widgets/FormatDate.tsx:
```diff
@@ -0,0 +1,101 @@
+
+    if (isRelative) {
+      // Dynamically updating display for client-side relative times
+      const interval = setInterval(update, 1000)
```
My suggestion would be to only kick of this interval if the time difference is less than an hour. After that it is unlikely these calls will result in a textual change.

acarapetis replied:
Good call, will do that.

robertvanhoesel commented on src/components/HypeCard.astro:
```diff
@@ -35,7 +35,7 @@ const website = (await getCollection("websites", ({ data }) => url?.includes(dat
         date && (
           <p>
-            <Timeago date={date} />
+            <FormatDate date={date} format="relative" />
```
```suggestion
            <FormatDate date={date} format="relative" client:load />
```

robertvanhoesel commented on src/components/widgets/FormatDate.tsx:
```diff
@@ -0,0 +1,101 @@
+    update()
+  })
+
+  return <>{output()}</>
```
Perhaps would be nice for relative dates to print the normal formatted date into a tooltip

acarapetis replied:
Oops, Timeago had that and I forgot to port it over.

acarapetis committed:
Hydrate FormatDate on HypeCard

Co-authored-by: Robert van Hoesel <robertvh@hey.com>

acarapetis committed:
FormatDate: add tooltip, only update recent times

For absolute dates[time]s, the tooltip always just shows the UTC version
of the date. For relative, this is updated to show the local absolute
time on the client-side.

Also implemented Robert's suggestion of only setting up dynamic updates
for relative times that are <1 hour old.

acarapetis committed:
FormatDate: Fix typo in shortrelative format

"s" was showing up as "m"

acarapetis commented:
@robertvanhoesel Thanks for the review! I've pushed some changes following your advice and removed the debugging code.

robertvanhoesel reviewed (approved):
Thanks, great work!
